llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +135,170 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
   return V;
 }
 
+hash_code hashIntrinsic(const Value *V) {
+  const auto *Extract = dyn_cast<ExtractValueInst>(V);
+  assert(Extract && "Not an Extract");
+  const auto *IntrinsicCall =
+      dyn_cast<CallBase>(Extract->getAggregateOperand());
+  assert(IntrinsicCall && "Not an Intrinsic");
+  Intrinsic::ID ID = IntrinsicCall->getIntrinsicID();
+  // Intrinsic ID is part of the hash.
+  hash_code Hash = hash_value(ID);
+  unsigned InPtrIdx;
+  isAIEPtrAddIntrinsic(ID, InPtrIdx);
+  // Necessary operands to desambiguate.
+  for (unsigned Op = InPtrIdx + 1; Op < hashableAddIntrinsicOps(ID); Op++) {
+    Hash = hash_combine(Hash, hash_value(IntrinsicCall->getArgOperand(Op)));
+  }
+  return Hash;
+}
+
+static std::pair<int, short> trackIntrinsicChain(const Value *V) {
+  int CountIntrinsicCall = 0;
+  hash_code Hash(0);
+  if (!V->getType()->isPointerTy())
+    return {-1, 0};
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+
+    if (auto *GEP = dyn_cast<GEPOperator>(V)) {
+      V = GEP->getPointerOperand();
+    } else if (Operator::getOpcode(V) == Instruction::BitCast ||
+               Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+      V = cast<Operator>(V)->getOperand(0);
+      if (!V->getType()->isPointerTy())
+        break;
+    } else {
+      if (auto *PHI = dyn_cast<PHINode>(V)) {
+        // Reached final destination;
+        return {CountIntrinsicCall, Hash};
+      } else if (auto *AIEObject = lookThroughAIEAddressingModes(V)) {
+        hash_code NewHash = hashIntrinsic(V);
+        // Initialize or compare hash?
+        if (Hash == hash_code(0)) {
+          Hash = NewHash;
+        } else if (Hash != NewHash) {
```

Comment:
It's not clear to me whether this hash compare could give false positives

Reply:
Sure! 16 bits is not too much! I will replace by a full set of operands!

Reply:
Done!